src/core/egg-herbie.rkt

```diff
@@ -182,12 +200,25 @@
     [3 "unsound"]
     [sr (error 'egraph-stop-reason "unexpected stop reason ~a" sr)]))
 
-;; An egraph is just a S-expr of the form
-;;  egraph ::= (<eclass> ...)
-;;  eclass ::= (<id> <enode> ..+)
-;;  enode  ::= (<op> <id> ...)
-(define (egraph-serialize egraph-data)
-  (egraph_serialize (egraph-data-egraph-pointer egraph-data)))
+;; Extracts the eclasses of an e-graph as a u32vector
+(define (egraph-eclasses egraph-data)
+  (egraph_get_eclasses (egraph-data-egraph-pointer egraph-data)))
+
+;; Extracts the nodes of an e-class as a vector
+;; where each enode is either a symbol, number, or list
+(define (egraph-get-eclass egraph-data id)
+  (define ptr (egraph-data-egraph-pointer egraph-data))
+  (define egg->herbie (egraph-data-egg->herbie-dict egraph-data))
+  (define eclass (egraph_get_eclass ptr id))
+  (for ([enode (in-vector eclass)]
+        [i (in-naturals)])
+    (match enode
+      [(cons (? number? n) _) (vector-set! eclass i n)] ; number
+      [(cons (? symbol? x) (? u32vector-empty?))
+       (when (hash-has-key? egg->herbie x) ; variable
+         (vector-set! eclass i x))]
+      [_ (void)])) ; everything else
```

Comment:
Up to you but I'd put this in egg-herbie.rkt probably, as a wrapper around `egraph-get-node`? No strong feelings.

Reply:
The interpretation of what is a variable and what is a constant is Herbie specific, so I'm more okay with it being here.

Reply:
Ok